### DISCUSSION TITLE
ui.timer invoking callback at startup

### DISCUSSION THREAD

**Discussion by valankar** [zauberzeug/nicegui#3742] [Q&A]
### Question

I noticed when I use ui.timer, my callback gets called immediately at startup, then it waits the elapsed time, then calls it again. Is this the expected behavior? I'd like to wait the elapsed time even at startup before invoking the first callback.

**Comment by rodja** [comment]
Yes, firing directly after the client connected is the intended behaviour. There is currently no parameter to change this. Would you like to create a pull request. I can imagine there are use cases where you want the execution to only be performed after the time elapsed.

**Reply by valankar** [reply]
Sure I will give it a shot.

**Reply by valankar** [reply]
Something I just noticed is _run_once sleeps before invoking the callback:

https://github.com/zauberzeug/nicegui/blob/d6d4ecd924efa1c01a3f085d85f9c7eefbc588dc/nicegui/elements/timer.py#L65

But _run_in_loop does a sleep after invoking the callback:

https://github.com/zauberzeug/nicegui/blob/d6d4ecd924efa1c01a3f085d85f9c7eefbc588dc/nicegui/elements/timer.py#L82

That seems a bit odd. Is there a reason for the difference?

**Reply by rodja** [reply]
Indeed. This is odd. We should clarify this in the documentation.

**Reply by valankar** [reply]
Is there a particular example where you expect the timer to run before actually waiting the specified time? That is rather non-intuitive to me, and I can't imagine anyone would expect it called immediately. It just seems like a bug in the code, not documentation.

**Reply by falkoschindler** [reply]
It is the intended behavior. Somehow I thought it would be reasonable to run before sleeping. If you want to run an important action in a certain interval, you might not want to wait for its first execution and start immediately. But it might be a matter of perspective. And I just noticed that JavaScript's `setInterval` also sleeps first (unlike `ui.timer`).

To avoid breaking changes that are hard to spot, we should keep the default behavior unchanged. But we can make it configurable.
